src/uu/ls/src/url_escaper.rs

```diff
@@ -0,0 +1,145 @@
+use std::{
+    borrow::Cow,
+    convert::{TryFrom, TryInto},
+    ffi::OsStr,
+};
+
+use ascii::{AsciiChar, AsciiStr, AsciiString};
+/// this is logically enum ByteOrResult {Char(AsciiChar), Escape([AsciiChar; 2])}
+/// but we don't use that definition because rustc uses 3 bytes instead of 2 for it (even if we use a pair instead of the array)
+#[derive(Copy, Clone)]
+#[repr(C)]
+struct AsciiOrEscape(Option<AsciiChar>, AsciiChar);
+
+impl AsciiOrEscape {
+    pub fn ascii(ch: AsciiChar) -> Self {
+        Self(None, ch)
+    }
+
+    pub fn escape(c0: AsciiChar, c1: AsciiChar) -> Self {
+        Self(Some(c0), c1)
+    }
+
+    #[inline]
+    pub fn classify(self) -> Result<AsciiChar, [AsciiChar; 2]> {
+        if let Some(c0) = self.0 {
+            Err([c0, self.1])
+        } else {
+            Ok(self.1)
+        }
+    }
+
+    pub fn is_byte(self) -> bool {
+        self.0.is_none()
+    }
+}
+
+pub struct UrlEscaper {
+    table: Box<[AsciiOrEscape; 256]>,
+}
+
+impl UrlEscaper {
+    pub fn new() -> Self {
+        let mut table = Vec::with_capacity(256);
+        static HEX_BYTES: &[u8; 16] = b"0123456789abcdef";
+        let hex_bytes = AsciiStr::from_ascii(HEX_BYTES).unwrap();
+        for &c0 in hex_bytes {
+            for &c1 in hex_bytes {
+                table.push(AsciiOrEscape::escape(c0, c1));
+            }
+        }
+        assert_eq!(table.len(), 256);
+        for r in [b'0'..=b'9', b'A'..=b'Z', b'a'..=b'z'] {
+            for b in r {
+                table[b as usize] = AsciiOrEscape::ascii(AsciiChar::from_ascii(b).unwrap());
+            }
+        }
+        for b in [b'_', b'-', b'.', b'~'] {
+            table[b as usize] = AsciiOrEscape::ascii(AsciiChar::from_ascii(b).unwrap());
+        }
+        Self {
+            table: table.into_boxed_slice().try_into().map_err(|_| ()).unwrap(),
```

Comment:
If we construct the table as a `Vec` and it's boxed anyway, why do we turn it into a `Box<[AsciiOrEscape; 256]>`? Is there a performance reason for that?